examples/multi-agent-debate/main.py

```diff
@@ -0,0 +1,325 @@
+        "waiting for user input?",
+        default=False,
+    )
+
+
+def run_debate() -> None:
+    """Execute the main debate logic.
+
+    Orchestrates the debate process, including setup, user input, LLM agent
+    interactions, and final feedback. Handles both user-guided and LLM-
+    delegated debates.
+
+    Raises:
+        Exception: If an unexpected error occurs during the debate process.
+    """
+    try:
+        # Get global settings
+        global_settings = get_global_settings(nocache=True)
+        langroid.utils.configuration.set_global(global_settings)
+
+        # Get base LLM configuration
+        agent_config: AgentConfig = get_base_llm_config()
+        system_messages: SystemMessages = load_system_messages(
+            "system_messages.json"
+        )
+        llm_delegate: bool = is_llm_delegate()
+
+        # Select topic and sides
+        selected_topic_tuple: Optional[tuple] = select_debate_topic(system_messages)
+        if not selected_topic_tuple:
+            logger.error("No topic selected. Exiting.")
+            return
+        topic_name, pro_key, con_key = selected_topic_tuple
+        side: str = select_side(topic_name)
+
+        # Create agents for pro, con, and feedback
+        pro_agent: Agent = create_agent(
+            agent_config, system_messages.messages[pro_key].message
+        )
+        con_agent: Agent = create_agent(
+            agent_config, system_messages.messages[con_key].message
+        )
+        feedback_agent: Agent = create_agent(
+            agent_config, system_messages.messages["feedback"].message
+        )
+        logger.info("Pro, Con, and feedback agents started.")
+
+        # Determine user's side
+        if side == "pro":
+            user_agent, ai_agent = pro_agent, con_agent
+            user_side, ai_side = "Pro", "Con"
+        else:
+            user_agent, ai_agent = con_agent, pro_agent
+            user_side, ai_side = "Con", "Pro"
+
+        logger.info(
+            f"Starting debate on topic: {topic_name}, taking the {user_side} side. "
+            f"LLM Delegate: {llm_delegate}"
+        )
+
+        # Prompt for the number of debate turns
+        max_turns: int = int(
+            Prompt.ask(
+                "How many turns should the debate continue for?",
+                default="4",
+            )
+        )
+        student_arguments: List[str] = []
+        ai_arguments: List[str] = []
+        is_user_turn: bool = True
+
+        for turn in range(max_turns):
+            if llm_delegate:
+                # Autonomous LLM debate
+                current_agent = user_agent if is_user_turn else ai_agent
+                agent_role = user_side if is_user_turn else ai_side
+                opponent_arguments = (
+                    ai_arguments if is_user_turn else student_arguments
+                )
+                context = (
+                    "\n".join(opponent_arguments[-1:])
```

Comment:
Here and elsewhere I think we can avoid this manual accumulation of conversation. E.g., we have the following input-output sequence of  msgs:

```
Student (or agent proxy) (StudentSysMsg) =>  S1  (i.e. student sees sys msg, says "S1")

AI (AI_Sys_msg, S1) =>  A1 (i.e. AI sees AI_Sys_Msg, S1, responds with A1)

Student (StudentSysMsg, S1, A1) => S2

AI(AI_SysMsg, S1, A1, S2) => A2
...
```

At each turn, each side is responding based on their own system msg, and **every** intervening msg of both sides, so you can simply pass on the `llm_response` of each agent to the other. I suggest as a simple exercise just set up two agents and try this. In fact we could probably create a task around each agent and have one task be subtask of the other, and count the turns and exit when needed. I'll see if I can mock up a simple example in next few days. You can try as well. I don't know if I'm over-simplifying what you're trying to do, but it will be nice to leverage langroid's built-in features if possible. I'll have to give it a closer look.

Reply:
For example you could probably re-work the teacher/student example in the langroid tour into a debate example: https://langroid.github.io/langroid/tutorials/langroid-tour/
The twist in what you're doing is the final feedback, and for that I think we could pass on a digest of the final responder's msg history to the feedback agent.